src/psyclone/psyir/nodes/dynamic_omp_task_directive.py

```diff
@@ -0,0 +1,1998 @@
+            elif isinstance(index, Reference):
+                index_private = self._is_reference_private(index)
+                # Check whether the reference is to a child loop variable.
+                child_loop_vars = []
+                for child_loop in self.walk(Loop):
+                    if child_loop.variable not in self._proxy_loop_vars:
+                        child_loop_vars.append(child_loop.variable)
+
+                if index_private:
+                    if (
+                        index not in private_list
+                        and index not in firstprivate_list
+                    ):
+                        firstprivate_list.append(index.copy())
+                    # Special case 1. If index belongs to a child loop
+                    # that is NOT a proxy for a parent loop, then we
+                    # can only do as well as guessing the entire range
+                    # of the loop is used.
+                    if index.symbol in child_loop_vars:
+                        # Return a :
+                        one = Literal(str(dim + 1), INTEGER_TYPE)
+                        lbound = BinaryOperation.create(
+                            BinaryOperation.Operator.LBOUND,
+                            Reference(ref.symbol),
+                            one.copy(),
+                        )
+                        ubound = BinaryOperation.create(
+                            BinaryOperation.Operator.UBOUND,
+                            Reference(ref.symbol),
+                            one.copy(),
+                        )
+                        full_range = Range.create(lbound, ubound)
+                        index_list.append(full_range)
+                    elif index.symbol in self._proxy_loop_vars:
+                        # Special case 2. the index is a proxy for a parent
+                        # loop's variable. In this case, we add a reference to
+                        # the parent loop's value. We create a list of all
+                        # possible variants, as we might have multiple values
+                        # set for a value, e.g. for a boundary condition if
+                        # statement.
+                        if len(index_list) <= dim:
+                            index_list.append([])
+                        for temp_ref in self._proxy_loop_vars[index.symbol][
+                            "parent_node"
+                        ]:
+                            parent_ref = temp_ref.copy()
+                            if isinstance(parent_ref, BinaryOperation):
+                                quick_list = []
+                                self._handle_index_binop(
+                                    parent_ref,
+                                    quick_list,
+                                    firstprivate_list,
+                                    private_list,
+                                )
+                                for element in quick_list:
```

Comment:
This could be a one-liner but ideally this whole method needs refactoring to reduce duplication.